src/ctia/lib/async.clj

```diff
@@ -63,28 +77,40 @@
    ec - An event channel, created with new-event-channel.
    f - user provided function that will be called when an event arrives on the event channel.
    pred - a predicate to test events. The user function will only be run if this predicate passes.
-   shutdown-fn - an optional function to run at system shutdown time. May be nil."
-  [{m :mult :as ec} :- EventChannel
-   listener-fn :- (=> s/Any Event)
-   pred :- (=> s/Bool Event)
-   shutdown-fn :- (s/maybe (=> s/Any))]
-  (let [events (chan)]
-    (tap m events)
-    (let [end-chan (chan)]
-      (go-loop []
-        (if-let [event (alt! [events end-chan] ([v] v))]
-          (do
-            (when (pred event)
-              (listener-fn event))
-            (recur))
-          (do
-            (close! end-chan)
-            (close! events))))
-      (.addShutdownHook (Runtime/getRuntime)
-                        (Thread. #(do (a/close! end-chan)
-                                      (when (fn? shutdown-fn)
-                                        (shutdown-fn)))))
-      end-chan)))
+   shutdown-fn - an optional function to run at system shutdown time. May be nil.
+   mode - can be :compute (for a green thread) or :blocking (for a real thread)"
+  ([ec :- EventChannel
+    listener-fn :- (=> s/Any Event)
+    pred :- (=> s/Bool Event)
+    shutdown-fn :- (s/maybe (=> s/Any))]
+   (register-listener ec listener-fn pred shutdown-fn :compute))
+  ([{m :mult :as ec} :- EventChannel
+    listener-fn :- (=> s/Any Event)
+    pred :- (=> s/Bool Event)
+    shutdown-fn :- (s/maybe (=> s/Any))
+    mode :- (s/enum :compute :blocking)]
+   (let [events (chan)
+         end-chan (chan)]
+     (tap m events)
+     (case mode
+       :blocking
+       (listen :mode :blocking
+               :pred pred
+               :listener-fn listener-fn
+               :events events
+               :end-chan end-chan)
+
+       :compute
+       (listen :mode :compute
+               :pred pred
+               :listener-fn listener-fn
+               :events events
+               :end-chan end-chan))
```

Comment:
Since listen is a macro, it needs to be passed the mode at compile time.  It subtle.  :)